xarray/core/variable.py

```diff
@@ -932,11 +936,7 @@ def copy(self, deep=True, data=None):
                 # don't share caching between copies
                 data = indexing.MemoryCachedArray(data.array)
 
-            if deep and (
-                hasattr(data, "__array_function__")
-                or isinstance(data, dask_array_type)
-                or (not IS_NEP18_ACTIVE and isinstance(data, np.ndarray))
-            ):
+            if deep and is_duck_array(data):
                 data = copy.deepcopy(data)
```

Comment:
I realize that it was removed in #4379 and not here, but do we need the special casing for `PandasIndexingAdapter`?

Reply:
Is that covered by this test?

https://github.com/pydata/xarray/blob/13caf96efb3f121e232a35aafceed80c832a9876/xarray/tests/test_variable.py#L512-L521

Reply:
unfortunately not, the special case is for when `data` is neither a duck array (or a duck dask array / a numpy array) nor a `PandasIndexAdapter`, but I'm not sure what the data could be instead. Maybe a `MemoryCachedArray`?

Reply:
actually, after changing
```python
            if deep and is_duck_array(data):
                data = copy.deepcopy(data)
```
to
```python
            if deep:
                if is_duck_array(data):
                    data = copy.deepcopy(data)
                elif not isinstance(data, PandasIndexAdapter):
                    raise ValueError(f"not a index adapter: {data}, {type(data)}")
```
does not cause any test in the test suite fail, which suggests that either the test suite is lacking (or the error is suppressed somehow) or this is a piece of obsolete code. Any insights on this, @shoyer?

Reply:
The main case where non-duck-arrays can be found in `_data` is when the data was loaded from a file on disk.

My inclination is that we should simplify this and probably use a deep copy even for non-duckarray objects. This might do a few extra copies sometimes, but it's better to be overly aggressive when doing a "deep" copy.

For example, even pandas.Index itself is only "mostly" immutable. You can't change the data values, but you can modify  `.name`.